src/scalar_weighting.rs

```diff
@@ -0,0 +1,232 @@
+        layer_dropout_prob: f64,
+    ) -> Self {
+        assert!(
+            n_labels > 0,
+            "The number of labels ({}) should be larger than 0",
+            n_labels
+        );
+
+        assert!(
+            n_features > 0,
+            "The number of features ({}) should be larger than 0",
+            n_features
+        );
+
+        let vs = vs.borrow();
+
+        let ws = vs.var("weight", &[n_labels, n_features], Init::KaimingUniform);
+        let bs = vs.var("bias", &[n_labels], Init::Const(0.));
+
+        ScalarWeightClassifier {
+            scalar_weight: ScalarWeight::new(vs.sub("scalar_weight"), n_layers, layer_dropout_prob),
+            linear: Linear { ws, bs },
+        }
+    }
+
+    pub fn forward(&self, layers: &[impl LayerOutput], train: bool) -> Tensor {
+        let logits = self.logits(layers, train);
+        logits.softmax(-1, Kind::Float)
+    }
+
+    pub fn logits(&self, layers: &[impl LayerOutput], train: bool) -> Tensor {
+        let features = self.scalar_weight.forward(layers, train);
+        self.linear.forward(&features)
+    }
+
+    /// Compute the losses and correctly predicted labels of the given targets.
+    ///
+    /// `targets` should be of the shape `[batch_size, seq_len]`.
+    pub fn losses(
+        &self,
+        layers: &[impl LayerOutput],
+        targets: &Tensor,
+        label_smoothing: Option<f64>,
+        train: bool,
+    ) -> (Tensor, Tensor) {
+        let targets_shape = targets.size();
+        let batch_size = targets_shape[0];
+        let seq_len = targets_shape[1];
+
+        let n_labels = self.linear.ws.size()[0];
+
+        let logits = self
+            .logits(layers, train)
+            .view([batch_size * seq_len, n_labels]);
+        let targets = targets.view([batch_size * seq_len]);
+
+        let predicted = logits.argmax(-1, false);
+
+        let losses =
+            cross_entropy_loss(&logits, &targets, self.linear.bs.size()[0], label_smoothing)
+                .view([batch_size, seq_len]);
+
+        (losses, predicted.eq1(&targets).view([batch_size, seq_len]))
+    }
+}
+
+fn cross_entropy_loss(
+    logits: &Tensor,
+    targets: &Tensor,
+    n_classes: i64,
+    label_smoothing: Option<f64>,
+) -> Tensor {
+    let probs = logits.log_softmax(-1, Kind::Float);
+
+    match label_smoothing {
+        Some(label_smoothing) => {
+            // Set all labels to label_smoothing and the target to 1-label_smoothing.
+            let n_classes = n_classes;
+            let smoothed_targets = tch::no_grad(|| {
+                Tensor::full_like(&probs, label_smoothing / (n_classes - 1) as f64).scatter(
+                    1,
+                    &targets.unsqueeze(1),
+                    &(1. - label_smoothing).into(),
+                )
+            });
+            (-smoothed_targets * probs).sum1(&[-1], false, Kind::Float)
+        }
+        None => probs.g_nll_loss::<&Tensor>(&targets, None, Reduction::None, -100),
```

Comment:
Will the losses be reduced at another point?

Reply:
Indeed, but this layer does not know anything about masking, so the losses are given as-is.